### PR TITLE
Update Discourse and Discourse plugins

### PR DESCRIPTION
**Pull request by MTRNord**
## Description of changes

This updates discourse to the latest available tag using the documented update process. Both https://meta.discourse.org/t/discourse-version-3-1/239039/2 (in parts since we went from the latest beta version to stable to 3.1.1 and then 3.2.0.beta1 in this PR) and https://meta.discourse.org/t/discourse-version-3-2/278584 are relevant.

Note that the 3.1 changes that came after the beta4 also included further security fixes. (I noticed "Don’t reuse CSP nonce between requests" for example to not be in 3.1.0.beta4 but in 3.1.1)

Noteworthy is the introduction of esbuild. This needed a new patch to work otherwise it cant find esbuild.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SebTM** [comment]
Maybe it makes sense to introduce a package-structure like signal-desktop with a beta-package to move faster but letting the user decide features vs. stability? This could also make back-porting easier as things can be tested earlier.

**Comment by MTRNord** [comment]
> Maybe it makes sense to introduce a package-structure like [signal-desktop](https://github.com/NixOS/nixpkgs/tree/nixos-unstable/pkgs/applications/networking/instant-messengers/signal-desktop) with a beta-package to move faster but letting the user decide features vs. stability? This could also make back-porting easier as things can be tested earlier.

I am torn. On the one hand I see what you mean and agree but on the other hand according to https://meta.discourse.org/t/need-a-better-way-to-explain-what-branch-to-be-on-why-and-what-happens/240342 the [test-passed](https://github.com/discourse/discourse/tree/tests-passed) branch is actually used in production at many pages. And the beta tags afaik are cut from that branch. So weirdly enough, it's considered stable enough that discourse and pages like ansible run it in prod.

So tldr: If we go by how it is run in the wild, the beta stuff seems to generally be agreed upon as stable by people running this at scale.

**Comment by drupol** [comment]
Hello,

This morning I tried to upgrade my discourse instance on our NixOS server and it failed:

```
❯ deploy .#cryptpad-nix
🚀 ℹ [deploy] [INFO] Running checks for flake in .
trace: warning: 
      `services.postgresql.ensureUsers.*.ensurePermissions` is used in your expressions,
      this option is known to be broken with newer PostgreSQL versions,
      consider migrating to `services.postgresql.ensureUsers.*.ensureDBOwnership` or
      consult the release notes or manual for more migration guidelines.

      This option will be removed in NixOS 24.05 unless it sees significant
      maintenance improvements.
    
warning: unknown flake output 'deploy'
error: builder for '/nix/store/c3m2dl9pmx8qfmiglgj4cgzn1yvp44ma-discourse-assets-3.2.0.beta1.drv' failed with exit code 1;
       last 25 log lines:
       > <internal:/nix/store/4khh1qa69fcca0h3fyqqwak963capr4v-ruby-3.2.2/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:38:in `require'
       > <internal:/nix/store/4khh1qa69fcca0h3fyqqwak963capr4v-ruby-3.2.2/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:38:in `require'
       > /nix/store/64c4nzqfyv81h7dlm7wxcj98jxxkjfz4-ruby3.2.2-bootsnap-1.16.0/lib/ruby/gems/3.2.0/gems/bootsnap-1.16.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
       > /nix/store/xbdfvgvcljyly5gq1qi83przdwv9r98d-ruby3.2.2-zeitwerk-2.6.11/lib/ruby/gems/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/kernel.rb:38:in `require'
       > /nix/store/dz1jrmh0bjzlppsj672dn7bdvipgnz0v-ruby3.2.2-railties-7.0.7/lib/ruby/gems/3.2.0/gems/railties-7.0.7/lib/rails/application.rb:348:in `require_environment!'
       > /nix/store/dz1jrmh0bjzlppsj672dn7bdvipgnz0v-ruby3.2.2-railties-7.0.7/lib/ruby/gems/3.2.0/gems/railties-7.0.7/lib/rails/application.rb:506:in `block in run_tasks_blocks'
       > /nix/store/q4r1gmp87dr4bz9w0n3fpndpk38p0qa6-ruby3.2.2-sprockets-rails-3.4.2/lib/ruby/gems/3.2.0/gems/sprockets-rails-3.4.2/lib/sprockets/rails/task.rb:61:in `block (2 levels) in define'
       > /nix/store/c8zwvx1i9m5a7q7w6qgmjc65mvmpmwc9-discourse-ruby-env-3.2.0.beta1/lib/ruby/gems/3.2.0/gems/rake-13.0.6/exe/rake:27:in `<top (required)>'
       > /nix/store/c5yn7x0ackjhb6ik7hwad100bhgwk4xk-bundler-2.4.22/lib/ruby/gems/3.2.0/gems/bundler-2.4.22/lib/bundler/cli/exec.rb:58:in `load'
       > /nix/store/c5yn7x0ackjhb6ik7hwad100bhgwk4xk-bundler-2.4.22/lib/ruby/gems/3.2.0/gems/bundler-2.4.22/lib/bundler/cli/exec.rb:58:in `kernel_load'
       > /nix/store/c5yn7x0ackjhb6ik7hwad100bhgwk4xk-bundler-2.4.22/lib/ruby/gems/3.2.0/gems/bundler-2.4.22/lib/bundler/cli/exec.rb:23:in `run'
       > /nix/store/c5yn7x0ackjhb6ik7hwad100bhgwk4xk-bundler-2.4.22/lib/ruby/gems/3.2.0/gems/bundler-2.4.22/lib/bundler/cli.rb:492:in `exec'
       > /nix/store/c5yn7x0ackjhb6ik7hwad100bhgwk4xk-bundler-2.4.22/lib/ruby/gems/3.2.0/gems/bundler-2.4.22/lib/bundler/vendor/thor/lib/thor/command.rb:28:in `run'
       > /nix/store/c5yn7x0ackjhb6ik7hwad100bhgwk4xk-bundler-2.4.22/lib/ruby/gems/3.2.0/gems/bundler-2.4.22/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
       > /nix/store/c5yn7x0ackjhb6ik7hwad100bhgwk4xk-bundler-2.4.22/lib/ruby/gems/3.2.0/gems/bundler-2.4.22/lib/bundler/vendor/thor/lib/thor.rb:527:in `dispatch'
       > /nix/store/c5yn7x0ackjhb6ik7hwad100bhgwk4xk-bundler-2.4.22/lib/ruby/gems/3.2.0/gems/bundler-2.4.22/lib/bundler/cli.rb:34:in `dispatch'
       > /nix/store/c5yn7x0ackjhb6ik7hwad100bhgwk4xk-bundler-2.4.22/lib/ruby/gems/3.2.0/gems/bundler-2.4.22/lib/bundler/vendor/thor/lib/thor/base.rb:584:in `start'
       > /nix/store/c5yn7x0ackjhb6ik7hwad100bhgwk4xk-bundler-2.4.22/lib/ruby/gems/3.2.0/gems/bundler-2.4.22/lib/bundler/cli.rb:28:in `start'
       > /nix/store/c8zwvx1i9m5a7q7w6qgmjc65mvmpmwc9-discourse-ruby-env-3.2.0.beta1/lib/ruby/gems/3.2.0/gems/bundler-2.4.22/exe/bundle:37:in `block in <top (required)>'
       > /nix/store/c5yn7x0ackjhb6ik7hwad100bhgwk4xk-bundler-2.4.22/lib/ruby/gems/3.2.0/gems/bundler-2.4.22/lib/bundler/friendly_errors.rb:117:in `with_friendly_errors'
       > /nix/store/c8zwvx1i9m5a7q7w6qgmjc65mvmpmwc9-discourse-ruby-env-3.2.0.beta1/lib/ruby/gems/3.2.0/gems/bundler-2.4.22/exe/bundle:29:in `<top (required)>'
       > /nix/store/4khh1qa69fcca0h3fyqqwak963capr4v-ruby-3.2.2/bin/bundle:25:in `load'
       > /nix/store/4khh1qa69fcca0h3fyqqwak963capr4v-ruby-3.2.2/bin/bundle:25:in `<main>'
       > Tasks: TOP => environment
       > (See full trace by running task with --trace)
       For full logs, run 'nix log /nix/store/c3m2dl9pmx8qfmiglgj4cgzn1yvp44ma-discourse-assets-3.2.0.beta1.drv'.
error: 1 dependencies of derivation '/nix/store/swn12p1jh56q6zs0kf9xag7mhgyrf6f7-discourse-3.2.0.beta1.drv' failed to build
error: 1 dependencies of derivation '/nix/store/0cv2b176k7bd138wnf0lk3x1d4kgpzm8-discourse-rake.drv' failed to build
error: 1 dependencies of derivation '/nix/store/4jxriw8s4dic3gn5zf0l3chddaqamxyl-nginx.conf.drv' failed to build
error: 1 dependencies of derivation '/nix/store/v4rk32d49ixqw8qhlv1lsxvxdfd4r2hq-unit-discourse.service.drv' failed to build
error: 1 dependencies of derivation '/nix/store/67skks1g7ds2rqlhlcg3gr3qsbj5n18k-discourse-rake_fish-completions.drv' failed to build
error: 1 dependencies of derivation '/nix/store/b9d0kh1niqs2myqaixkrh7ii8sr2sdi3-man-paths.drv' failed to build
error: 1 dependencies of derivation '/nix/store/jvcjnawggg56v22930gl3cih0wsnfs2c-system-path.drv' failed to build
error: 1 dependencies of derivation '/nix/store/hi87masfq9n8a8wzhzvwf30y4490m5jg-system-units.drv' failed to build
error: 1 dependencies of derivation '/nix/store/zlr9vin7lj4j10gzvbnirwnlj20am8s2-nixos-system-discourse-24.05.20231217.91a0070.drv' failed to build
error: 1 dependencies of derivation '/nix/store/c0da26i176ynz823w6jqjvzcc2aq9sml-activatable-nixos-system-discourse-24.05.20231217.91a0070.drv' failed to build
error: 1 dependencies of derivation '/nix/store/611rjamvdkszlx2vmj3sm6dssapklfa8-deploy-rs-check-activate.drv' failed to build
error: build of '/nix/store/611rjamvdkszlx2vmj3sm6dssapklfa8-deploy-rs-check-activate.drv', '/nix/store/ki452wkvf5i0qrgpnj5n7wakg16g3n4v-jsonschema-deploy-system.drv' failed
🚀 ❌ [deploy] [ERROR] Failed to check deployment: Nix checking command resulted in a bad exit code: Some(1)
```

Then I checked the latest tag from `discourse/discourse`: https://github.com/discourse/discourse/tags and `3.2.0beta3` is the latest one.

Therefore, this commit message is a bit missleading since it is: `[discourse: 3.1.0.beta4 -> 3.2.0.beta4](https://github.com/NixOS/nixpkgs/pull/257294/commits/beee2842bdc9281c94c95ea9c89f20b3da53ffd3)`
This is not an upgrade to `3.2.0beta4` but to `3.2.0beta1`.

Also, do you think we could plan an upgrade of Discourse soon ?

**Comment by MTRNord** [comment]
The commit is a typo. I only realised this weeks after this was merged :/ sorry for that.

I am not sure about the error. I am using this on 2 nixos servers of which both are flake based.

**Comment by drupol** [comment]
Yeah, my server is also flake based and updating the flake is breaking discourse :(

**Comment by drupol** [comment]
Is there any plan to upgrade Discourse to version 3.2.0 ?
